impeller/core/vertex_buffer.h

```diff
@@ -5,13 +5,14 @@
 #ifndef FLUTTER_IMPELLER_CORE_VERTEX_BUFFER_H_
 #define FLUTTER_IMPELLER_CORE_VERTEX_BUFFER_H_
 
+#include <variant>
 #include "impeller/core/buffer_view.h"
 #include "impeller/core/formats.h"
 
 namespace impeller {
 
 struct VertexBuffer {
-  BufferView vertex_buffer;
+  std::variant<BufferView, std::vector<BufferView>> vertex_buffers;
```

Comment:
This is a bit of a code smell to me. Pretend the VertexBuffer struct doesn't exist. The render pass API would probably use overloads, something like:

```
pass.setVertexBuffer(BufferView vertex_buffer);
pass.setVertexBuffer(BufferView index_buffer, IndexType type, BufferView vertex_buffer);
pass.setVertexBuffer(BufferView index_buffer, IndexType type, BufferView[] vertex_buffers)

```

what I think this change is doing is essentially moving the overload inside of the struct with the variant. Arguably this struct isn't even necessary.

Binding multiple vertex buffers should be done through a different API that accepts N vertex buffers instead of updating the VertexBuffer struct. That API could also pass them via iterator as to avoid more heap allocations for Vulkan/Metal which can immediately encode them.

Reply:
If you set a reasonable limit to the number of bound vertex buffers too, then you can have the various render pass structs hold them without doing a heap allocation per draw.

Reply:
That's a good point. For SoA layouts, we'd need one BufferView per vertex attribute. Looking around, maybe a reasonable limit would be 16.
- For Metal, Apple GPUs are limited to 31 vertex attributes ([Maximum number of vertex attributes, per vertex
descriptor](https://developer.apple.com/metal/Metal-Feature-Set-Tables.pdf))
- GL implementations expose the attribute limit through `GL_MAX_VERTEX_ATTRIBS`, and according to [this](https://www.khronos.org/opengl/wiki/vertex_Specification#Separate_attribute_format): "This will almost certainly be 16."
- Vulkan implementations expose the attribute limit through `vkGetPhysicalDeviceProperties` ([VkPhysicalDeviceLimits.maxVertexInputBindings](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkPhysicalDeviceLimits.html)). [wgpu](https://github.com/gfx-rs/wgpu/blob/fe7fbd4a4abd0d9e8269c9df5e1dd3f8846d6462/wgpu-hal/src/lib.rs#L295) hardcodes their vertex buffer limit to 16.

The consensus seems to be that virtually everything supports 16 and more modern hardware supports ~32. 16 isn't even enough for glTFs with morph targets, but it's a bit ridiculous that those aren't interleaved anyway IMO...

Reply:
I think the ideal layout varies per target platform as well. my limited understanding is that for tilers its advantageous to have position data in a separate buffer and all other varyings interleaved.

Reply:
I changed this to add an API for setting the VertexBuffer and IndexBuffer config directly on `RenderPass`.

I think we should dissolve `VertexBuffer` completely, but we have a pattern across the codebase where we pass around `GeometryResult` structs with vertex buffers. I think the right move here would be to switch stuff like `GetPositionBuffer` to `BindPositionBuffer` and immediately bind results rather than deferring the bind to the parent scope.

That's a bigger change. I could do it as an immediate follow-up, but I can also do that now if you'd prefer.

Reply:
I agree with removing VertexBuffer eventually. I think I would leave VertexBuffer as is, and instead change flutter_gpu to not use it.